### DISCUSSION TITLE
Add ETHERNET support (W5500)

### DISCUSSION THREAD

**Discussion by MicSG-dev** [mathieucarbou/ESPAsyncWebServer#36] [Ideas]
It would be great to be able to use an ethernet module like the W5500 on the ESP32 and be able to use this library.

![image](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71986598/69a93ef2-a7aa-44cb-a25b-6704ef5a9acd)

**Comment by mathieucarbou** [comment]
@MicSG-dev This lib is a web server, it has nothing to do with which interface is used or not.

FYI, I am using w32-eth and lilygo eth boards with success with [MycilaESPConnect](https://github.com/mathieucarbou/MycilaESPConnect), this ESPAsyncWebServer lib and ESP-DASH for example.

**Reply by MicSG-dev** [reply]
Ok. Could you show me an example of using the ESP32 with ethernet with the ESPAsyncWebServer library?

**Reply by mathieucarbou** [reply]
Look at the examples in [MycilaESPConnect](https://github.com/mathieucarbou/MycilaESPConnect). [MycilaESPConnect](https://github.com/mathieucarbou/MycilaESPConnect) is using ESPAsyncWebServer...

**Reply by MicSG-dev** [reply]
I looked at all the examples in the [MycilaESPConnect](https://github.com/mathieucarbou/MycilaESPConnect) library and none of them contained ethernet usage (wifi only). I searched for '[w5500](https://github.com/search?q=repo%3Amathieucarbou%2FMycilaESPConnect%20w5500&type=code)' in the library repository, and found some references to the W5500. But how to make use of it with the library?

**Reply by mathieucarbou** [reply]
you need to define the pinouts in the pio file for your ethernet card, like here: 
https://github.com/mathieucarbou/MycilaESPConnect/blob/main/platformio.ini#L111-L119
It is also written in the project's readme.

eth support is not linked to a library but instead how you wire your pins and setup the Eth library.

Read the code and take time to understand it if you need. But you have everything inside this lib. It shows you how to use Eth in Arduino 2 and 3 with and without SPI.

**Reply by MicSG-dev** [reply]
@mathieucarbou I managed to execute it🎉🎉!! Thanks for the guidance. The only problem I faced was with PlatformIO, due to the IDF version, which by default is 4.x. After some research, I was able to use IDF version 5.x on PlatformIO. Below are the files for PlatformIO:


- file platformio.ini:

````
; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[env:esp32dev]
platform = espressif32
platform_packages=
  platformio/framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#3.0.1
  platformio/framework-arduinoespressif32-libs @ https://github.com/espressif/arduino-esp32/releases/download/3.0.1/esp32-arduino-libs-3.0.1.zip
board = esp32dev
framework = arduino
build_flags = ${env.build_flags}
  -D ESPCONNECT_ETH_SUPPORT
  -D ETH_PHY_ADDR=1
  -D ETH_PHY_CS=5
  -D ETH_PHY_IRQ=4
  -D ETH_PHY_RST=14
  -D ETH_PHY_SPI_MISO=19
  -D ETH_PHY_SPI_MOSI=23
  -D ETH_PHY_SPI_SCK=18
  -D ETH_PHY_TYPE=ETH_PHY_W5500
lib_deps = 
    https://github.com/mathieucarbou/MycilaESPConnect.git
````

- file main.cpp:
````
#include <Arduino.h>
#include <MycilaESPConnect.h>

uint32_t lastLog = 0;

AsyncWebServer server(80);

void setup() {
  Serial.begin(115200);
  while (!Serial)
    continue;

    Serial.print("ESP_IDF_VERSION_MAJOR (it has to be 5 to work): ");
    Serial.println(ESP_IDF_VERSION_MAJOR);

  
  // serve your home page here
  server.on("/", HTTP_GET, [&](AsyncWebServerRequest* request) {
    request->send(200, "text/plain", "Hello World!");
  });


  ESPConnect.begin(server, "arduino", "Captive Portal");

  Serial.println("ESPConnect completed, continuing setup()...");

  server.begin();
}

void loop() {
  ESPConnect.loop();

  if (millis() - lastLog > 5000) {
    JsonDocument doc;
    ESPConnect.toJson(doc.to<JsonObject>());
    serializeJson(doc, Serial);
    Serial.println();
    lastLog = millis();
  }
}
````

- Pin Connection on ESP32:
    - MISO_PIN: 19
    - MOSI_PIN: 23
    - SCLK_PIN: 18
    - CS_PIN: 5
    - INT_PIN (IRQ): 4
    - RST_PIN: 14

**Reply by mathieucarbou** [reply]
Awesome!
~~Feel free to contribute a pull request to add an entry in the pio file the MycilaESPConnect.git project, in order to give other people an example of wiring for this ETH card. The readme could also be updated with the table of tested cards and a link.~~
Thanks!

**Reply by mathieucarbou** [reply]
Never mind: did it already ;-)
https://github.com/mathieucarbou/MycilaESPConnect?tab=readme-ov-file#ethernet-support

**Reply by MicSG-dev** [reply]
@mathieucarbou 👍. Thank you for your efforts in the Arduino community!

**Reply by MicSG-dev** [reply]
To make things easier, I made the schematics:
![w5500_lite_with_esp32_skematic1](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71986598/fc6a844e-9915-4ca7-86aa-4a75828968a4)

![w5500_lite_with_esp32_skematic2](https://github.com/mathieucarbou/ESPAsyncWebServer/assets/71986598/6065c746-124a-4ee8-a33b-eb434c6e1b0e)

*Note: I didn't use the RST pin, but it still worked.

**Comment by proddy** [comment]
The W5500 is more complex to implement (needs SPI and lots of pins) and not as stable as say the LAN8720 (which is a PHY chip) which is cheaper and more common.

**Reply by mathieucarbou** [reply]
FYi [MycilaESPConnect](https://github.com/mathieucarbou/MycilaESPConnect) supports SPI, even with Arduino 2. Example with Lilygo ETH board.

**Reply by egnor** [reply]
Also not every ESP32 variant has an Ethernet MAC [edit to add:] controller peripheral onboard with MII/RMII support for an external PHY. So either you're using a W5500 sorta thing or you're using an external MAC which is all kinds of extra fun...

**Reply by mathieucarbou** [reply]
> Also not every ESP32 variant has an Ethernet MAC onboard. So either you're using a W5500 sorta thing or you're using an external MAC which is all kinds of extra fun

ESP's are generating the mac address for the ETH layer based on an index added to the fuse mac address (if I'm not mistaken). So when this MAC address is not provided or to avoid a random generation, we can use the SPI to force a mac address that will be consistant in order to get the same mac address for the same device on both arduino 2 and 3.  I ran into that in MycilaESPConnect and I am handing this situation by incremented the fuse mac address.

_edit_

Actually, I had to do that to have a consistant mac address generation for lilygo eth board on esp32 s3 when on arduino 2 which does not support spi:

- read mac address as generated by the esp-idf api
- set it in the SPI impl

```
  uint8_t address[6] = {0, 0, 0, 0, 0, 0};
  if (esp_read_mac(address, ESP_MAC_ETH) != ESP_OK) {
    log_e("esp_read_mac failed");
    return false;
  }
  esp_eth_ioctl(eth_handle, ETH_CMD_S_MAC_ADDR, address);
```

**Reply by mathieucarbou** [reply]
@egnor : Oh ;-) Thanks a lot by the way for your awesome w32-eth repo (https://github.com/egnor/wt32-eth01) - it helped me! I hate this board lol ;-) Not fun to play wit (I don't have the dongle that would allow an automatic boot switch and reset)

**Reply by egnor** [reply]
Sorry, by MAC I wasn't referring to the MAC address, but to the Ethernet MAC (Media Access Control) implementation. (Confusing-- my bad!) A microcontroller can't just talk directly to a PHY without a special peripheral that implements [the MII or RMII (or for gigabit, GMII, etc) interface](https://en.m.wikipedia.org/wiki/Media-independent_interface) that a PHY chip offers.

The original ESP32 includes an onboard MAC peripheral and can talk directly to PHY chips. More recent variants (ESP32-S2, -C3, etc) often do not (not enough demand, they used the silicon for something else). So if you want to do wired Ethernet on those chips, you need discrete MAC and PHY chips (or a MAC+PHY combo), you can't just hook up a PHY to those chips like @proddy was advising.

**Reply by mathieucarbou** [reply]
> The original ESP32 includes an onboard MAC peripheral and can talk directly to PHY chips. More recent variants (ESP32-S2, -C3, etc) often do not (not enough demand, they used the silicon for something else

Thanks! didn't know that :-)

**Comment by egnor** [comment]
Clarification for the record (please correct any misconceptions I have here!): It's not like there's a single network layer in the Arduino type world, so it's not QUITE right to say that ESPAsyncWebServer is independent of the network driver.

The Arduino platform has a `Client` abstraction which is used in `WiFi` and `Ethernet`. Many network libraries provide this interface because that's what a lot of Arduino code uses. For example this is what [the TinyGSM cell modem library](https://github.com/vshymanskyy/TinyGSM) and [the Ethernet3 W5500 library](https://github.com/sstaub/Ethernet3) offer. But, ESPAsyncWebServer does NOT use Client/Ethernet/WiFi and will NOT work with networking libraries that provide those interfaces. You cannot (to the best of my knowledge) use ESPAsyncWebServer with, say, TinyGSM [edited] as its network provider (which is too bad!).

The ESP32 platform code (ESP-IDF) has its own network stack, based on [the lwip (Light Weight IP) library](https://github.com/lwip-tcpip/lwip) plus ESP32-specific hardware drivers. The arduino-esp32 implementation builds its Client/Ethernet/WiFi on top of that. ESPAsyncWebServer uses the lwip library interfaces directly (via AsyncTCP). So to make ESPAsyncWebServer work with a network interface you need a driver implementation that fits into the ESP-IDF network stack.

Fortunately ESP-IDF does include support for the W5500 over SPI, so by defining the right things you can make that work, as above. But Arduino-oriented W5500 libraries like Ethernet3 will not work.

**Reply by mathieucarbou** [reply]
> You cannot (to the best of my knowledge) use ESPAsyncWebServer with, say, TinyGSM (which is too bad!).

I am using both ESPAsyncWebServer and TinyGMS in a project (https://beelance.carbou.me) even with a GPS lib. Through this project we have access to the Onomondo platform (we made a deal with them). This project is to offer a remote scale for beehives through LTE-M / NB-IoT. It uses GPS, Modem (TinyGSM), ESPAsyncWebServer, etc. We also support normal sim cards.

So I guess that by "using" you mean for example have access to the internal ESPAsyncWebServer port through a modem com with TinyGSM ?

Because right now I saw no issue in using both and doing some remote calls through the modem, and this is considering both the SIM080G and the TA7670G are used, first one through the Arduino Client and the second one directly by using the AT commands and sending a binary stream to the modem because it does not have a Client implementation. But whatever the Client implementation is used or not this is always possible to directly communicate with the modem card through AT commands and send/receive bytes.

**Reply by egnor** [reply]
> So I guess that by "using" you mean for example have access to the internal ESPAsyncWebServer port through a modem com with TinyGSM ?

Oh yes. You can certainly have both libraries in one project and have them both running, but ESPAsyncWebServer can't use TinyGSM as a network layer.

(I assume in this project ESPAsyncWebServer serves over local WiFi for configuration during deployment / service in the field?)

**Reply by mathieucarbou** [reply]
Yes! I agree with you for the network layer. Yes we are not the modem to tunnel data from and to the devices in this project, but just to report. I didn't see anything in the AT spec of the DIM7080G and TA7670G anyway that would allow such use case ? All has to go through AT commands supporting sending bytes for different kind of protocols but to have the modem be used as a network layer, not sure we can with these ones.

**Reply by egnor** [reply]
I don't know about those two products but cell modems vary. They typically offer some mixture of built-in protocol-client support (via AT commands) and raw network packet access (via PPP). It's [semi standardized](https://www.etsi.org/deliver/etsi_ts/127000_127099/127007/15.02.00_60/ts_127007v150200p.pdf) though everyone has extensions of course.

TinyGSM will implement the Arduino "Client" interface on top of some of these by taking calls to `connect()` and turning them into AT commands that open a socket, so you're using the cell modem's TCP/IP stack rather than the host processor's. This is to some degree possible because "Client" is an extremely basic interface, it's blocking, non threaded, generally only lets you have one connection at a time, and so on.

In theeeeeory with some more advanced cell modems that allow multiple connections and multiplexing and stuff and some appropriate middle-layer interface you could target something like ESPAsyncWebServer to such an interface, but wow that would be a lot of work and the result would probably be buggy in a lot of ways. More plausibly you'd set up a PPP link and plumb that into the lwip stack, at which point ESPAsyncWebServer could just carry on using the interfaces it already does. But I don't actually know of any off-the-shelf software that does that.

Very few people are asking to serve HTTP off their cell modems anyway. What people DO want is MQTT, and there are a lot of issues there, but that's entirely outside the scope of this project!

**Reply by mathieucarbou** [reply]
Ah! you're right I only looked at the AT command datasheet ;-) But there are some others. I agree that it's not the main use case.

**Comment by juanmartin84** [comment]
Hello, i'm using a w5500
I follow almost the steps mentioned @MicSG-de with some pin variant because of my hardware.
In my case the interrupt pin is not connected, my hardware design does not allow this (maybe next revision). Is this a problem?
Below the platform.ini detail

```
build_flags = ${env.build_flags}
  -D ESPCONNECT_ETH_SUPPORT
  -D ETH_PHY_ADDR=1
  -D ETH_PHY_CS=5
  -D ETH_PHY_IRQ=4
  -D ETH_PHY_RST=26
  -D ETH_PHY_SPI_MISO=19
  -D ETH_PHY_SPI_MOSI=23
  -D ETH_PHY_SPI_SCK=18
```
I get this message on serial:

w5500.mac: emac_w5500_transmit(481): free size (350) < send length (0)

What does it mean?

**Reply by MicSG-dev** [reply]
> Hello, i'm using a w5500 I follow almost the steps mentioned @MicSG-de with some pin variant because of my hardware. In my case the interrupt pin is not connected, my hardware design does not allow this (maybe next revision). Is this a problem? Below the platform.ini detail
> 
> ```
> build_flags = ${env.build_flags}
>   -D ESPCONNECT_ETH_SUPPORT
>   -D ETH_PHY_ADDR=1
>   -D ETH_PHY_CS=5
>   -D ETH_PHY_IRQ=4
>   -D ETH_PHY_RST=26
>   -D ETH_PHY_SPI_MISO=19
>   -D ETH_PHY_SPI_MOSI=23
>   -D ETH_PHY_SPI_SCK=18
> ```
> 
> I get this message on serial:
> 
> w5500.mac: emac_w5500_transmit(481): free size (350) < send length (0)
> 
> What does it mean?

I saw that your platformio.ini file does not have the flag '-D ETH_PHY_TYPE=ETH_PHY_W5500'.
Try running a minimal example to make sure there is no interference from other hardware/libraries. Please show your complete platformio.ini file.

**Reply by MicSG-dev** [reply]
@juanmartin84 try running this basic example (it uses mathieucarbou's ESPAsyncWebServer library): https://github.com/MicSG-dev/Minimal-Example-ESP32-W5500-AsyncWebServer

**Reply by juanmartin84** [reply]
Hello, sorry for the delay
I get this on serial output

ETH Started
hostname: ESP32 123
mac: EE:62:60:FE:14:48
ETH Connected

But no ip is assigned, when I scan the network i can't see the device.
Is it working?

Just changed the .ini file like this

```
build_flags = ${env.build_flags}
  -D ESPCONNECT_ETH_SUPPORT
  -D ETH_PHY_ADDR=1
  -D ETH_PHY_CS=5
  -D ETH_PHY_IRQ=4
  -D ETH_PHY_RST=26
  -D ETH_PHY_SPI_MISO=19
  -D ETH_PHY_SPI_MOSI=23
  -D ETH_PHY_SPI_SCK=18
  -D ETH_PHY_TYPE=ETH_PHY_W5500
```

I can't change the pins (pcb is done), and I don't have connected ADDR  / interrupt pins.

**Reply by MicSG-dev** [reply]
> Hello, sorry for the delay I get this on serial output
> 
> ETH Started hostname: ESP32 123 mac: EE:62:60:FE:14:48 ETH Connected
> 
> But no ip is assigned, when I scan the network i can't see the device. Is it working?
> 
> Just changed the .ini file like this
> 
> ```
> build_flags = ${env.build_flags}
>   -D ESPCONNECT_ETH_SUPPORT
>   -D ETH_PHY_ADDR=1
>   -D ETH_PHY_CS=5
>   -D ETH_PHY_IRQ=4
>   -D ETH_PHY_RST=26
>   -D ETH_PHY_SPI_MISO=19
>   -D ETH_PHY_SPI_MOSI=23
>   -D ETH_PHY_SPI_SCK=18
>   -D ETH_PHY_TYPE=ETH_PHY_W5500
> ```
> 
> I can't change the pins (pcb is done), and I don't have connected ADDR / interrupt pins.

Hello.
Your device did not get an IP because it MUST have the interrupt pin connected (ETH_PHY_IRQ). This is because the <ETH.h> library requires the interrupt pin connected to manage events and there is nothing to be done if this pin is not connected. The ADDR you mentioned is not a pin, so there is no need to modify the hardware because of the ADDR. In your case, it is necessary to modify the hardware to use the interrupt pin.

My tip to fix the problem on your PCB is: solder a jumper wire between the ESP32 pin and the W5500 interrupt pin (it would be a workaround, but it is a solution :D).

If everything goes well, the serial monitor should print the message:
ETH got IP: "xxx.xxx.xxx.xx", where "xxx.xxx.xxx.xx" is the IP address obtained.

**Comment by juanmartin84** [comment]
Thanks @MicSG-dev ! 
I soldered an external wire with interrupt pin and worked as in your example.

Then I tried to integrate this in a bigger app but I get these errors, seems to be that the right ETH.h lib is not called

`
error: 'ETH_PHY_W5500' was not declared in this scope
error: 'Network' was not declared in this scope
error: 'onEvent' was not declared in this scope
error: 'SPI2_HOST' was not declared in this scope
error: 'ARDUINO_EVENT_ETH_LOST_IP' was not declared in this scope
`

here is my ini file:

```
[env:esp32dev]
platform = espressif32 @ 6.9.0
board = esp32dev
framework = arduino

monitor_speed = 115200
monitor_port = COM[34]; Windows, COM3 or COM4

board_build.flash_mode = dio
board_upload.flash_size = 4MB

board_build.f_cpu = 240000000L ;set clock frequency to 240MHz
board_build.f_flash = 80000000L ; set flash frequency to 80MHz
board_build.partitions = partition_table_HUGE_APP.csv ;https://github.com/espressif/arduino-esp32/tree/master/tools/partitions
board_build.filesystem = littlefs

upload_speed = 921600
upload_port = COM[34]; COM3 or COM4

lib_deps = 
  mathieucarbou/ESPAsyncWebServer@^3.3.16
  mathieucarbou/AsyncTCP@^3.2.10
  bblanchon/ArduinoJson@^7.1.0
  knolleary/PubSubClient@^2.8
  fbiego/ESP32Time@^2.0.6
  adafruit/Adafruit ADS1X15@^2.5.0

lib_compat_mode = strict
lib_ldf_mode = chain

monitor_filters = 
  esp32_exception_decoder
  time      ; Add timestamp with milliseconds for each new line

build_flags = ${env.build_flags}
  -D ESPCONNECT_ETH_SUPPORT
  -D ETH_PHY_TYPE=ETH_PHY_W5500
```

here the function:

```
AsyncWebServer server(84);

void setup_ethconnection(void) {
  
 Network.onEvent(onEvent);
  ETH.begin(ETH_PHY_TYPE, 1, 5, 12, 26, SPI2_HOST, 18, 19, 23);

  server.on("/", HTTP_GET, handleRootRequest);
  
  server.begin(); 

  }

void handleRootRequest(AsyncWebServerRequest *request)
{
  request->send(200, "text/plain", "Hello!");
}

void onEvent(arduino_event_id_t event, arduino_event_info_t info)
{
  switch (event)
  {
  case ARDUINO_EVENT_ETH_START:
  {
    Serial.println("ETH Started");
    const char *hostnameBoard = "ESP32 123";

    Serial.print("hostname: ");
    Serial.println(hostnameBoard);
    ETH.setHostname(hostnameBoard);

    Serial.print("mac: ");
    Serial.println(ETH.macAddress());

    break;
  }
  case ARDUINO_EVENT_ETH_CONNECTED:
  {
    Serial.println("ETH Connected");
    break;
  }
  case ARDUINO_EVENT_ETH_GOT_IP:

  {
    Serial.println("ETH got IP: " + ETH.localIP().toString());
    break;
  }
  case ARDUINO_EVENT_ETH_LOST_IP:
  {
    Serial.println("ETH Lost IP");
    break;
  }
  case ARDUINO_EVENT_ETH_DISCONNECTED:
  {
    Serial.println("ETH Disconnected");
    break;
  }
  case ARDUINO_EVENT_ETH_STOP:
  {
    Serial.println("ETH Stopped");
    break;
  }
  default:
  {
    break;
  }
  }
}
```

How can I fix it?
thanks

**Comment by juanmartin84** [comment]
Hello, I checek this lines in .ini file

If change  
platform = espressif32

by
platform = espressif32 @ 6.8.1

Then errors appear in the example that was working.

My problem is that using just "platform = espressif32" a lot of other errors appear.

**Reply by MicSG-dev** [reply]
> My problem is that using just "platform = espressif32" a lot of other errors appear.

Use 
(Don't use this anymore) ~"platform=espressif32"~
(Use this) `platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip`
 and show us the errors that occur.

**Comment by MicSG-dev** [comment]
Hello.
In your platformio.ini file try removing the version definition.

BEFORE
```
platform = espressif32 @ 6.9.0
```

AFTER
~platform = espressif32~ (Do not use this way anymore)
Use the 3.x package; In this case I'm using Arduino core 3.1.0 rc2 based on IDF 5.3.1+:
```
platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip
```


I did some tests here and from what I've seen this happens because when REMOVING the version in the .ini file, PlatformIO obtains the package from https://github.com/tasmota/arduino-esp32 (que É compatível com o W5500): 
![image](https://github.com/user-attachments/assets/a5da0e87-10f0-43d3-95ba-b497ec4399ea)

Note that the Ethernet/ETH_W5500_Arduino_SPI example is present: 
![image](https://github.com/user-attachments/assets/ebe4965c-d572-4074-8caa-932461c39f81)


By INSERTING version 6.9.0 into the .ini file, PlatformIO gets the package from https://github.com/platformio/platform-espressif32 (which is somehow NOT compatible with the W5500). Note that there is no Ethernet/ETH_W5500_Arduino_SPI example:
![image](https://github.com/user-attachments/assets/9fac4028-823c-40d7-b5d0-759aa668554d)


### The solution
~Remove the version information that is in the platformio.ini file. It should look like this:~
~platform = espressif32~

Use the 3.x package, where in this case below I'm using Arduino core 3.1.0 rc2 based on IDF 5.3.1+:
```
platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip
```

**Reply by mathieucarbou** [reply]
You should use pioarduno now with Arduino Core 3.x:

https://github.com/pioarduino/platform-espressif32

**Reply by MicSG-dev** [reply]
Thanks @mathieucarbou ! I was not aware of the impasse that exists/has existed between Espressif, PlatformIO and Arduino ESP32 Core version 3.0.0. I updated the [example](https://github.com/MicSG-dev/Minimal-Example-ESP32-W5500-AsyncWebServer) to use the latest package available: [Arduino core 3.1.0 rc2 based on IDF 5.3.1+](https://github.com/pioarduino/platform-espressif32/releases/tag/53.03.10-rc2)

@juanmartin84 The platformio.ini file will look like this:
```
platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip
```

**Comment by juanmartin84** [comment]
Hello @mathieucarbou @MicSG-dev thanks for the support.
Unfortunately after moving to Arduino Core 3.x i'm getting some errors during the run time, maybe this exceeds the ethernet topic.

I used :

`platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip
`
getting:

Backtrace: 0x4008354d:0x3ffb1fc0 0x4008e22d:0x3ffb1fe0 0x400945f6:0x3ffb2000 0x40134430:0x3ffb2130 0x4011de0a:0x3ffb2150 0x400e328a:0x3ffb21c0 0x400ef13a:0x3ffb2240 0x4011f6a3:0x3ffb2270 0x4008efaa:0x3ffb2290
22:17:31.205 >   #0  0x4008354d in panic_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/panic.c:463
22:17:31.205 >   #1  0x4008e22d in esp_system_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/port/esp_system_chip.c:92
22:17:31.205 >   #2  0x400945f6 in __assert_func at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/newlib/assert.c:80
22:17:31.205 >   #3  0x40134430 in sntp_setoperatingmode at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/lwip/lwip/src/apps/sntp/sntp.c:729 (discriminator 1)
22:17:31.205 >   #4  0x4011de0a in configTime at C:/Users/Juan/.platformio/packages/framework-arduinoespressif32/cores/esp32/esp32-hal-time.c:53
22:17:31.205 >   #5  0x400e328a in getDateTimeSetup() at src/dateTime.cpp:12
22:17:31.205 >   #6  0x400ef13a in setup() at src/main.cpp:159
22:17:31.205 >   #7  0x4011f6a3 in loopTask(void*) at C:/Users/Juan/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:59
22:17:31.205 >   #8  0x4008efaa in vPortTaskWrapper at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:139

and:

`platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip
`
getting:

22:06:53.761 >   #0  0x40131273 in memp_malloc at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/lwip/lwip/src/core/memp.c:370 (discriminator 2)
22:06:53.761 >   #1  0x4012f9d3 in tcpip_inpkt at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/lwip/lwip/src/api/tcpip.c:254 (discriminator 2)
22:06:53.761 >   #2  0x4012fa19 in tcpip_input at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/lwip/lwip/src/api/tcpip.c:290
22:06:53.761 >   #3  0x401493c9 in wlanif_input at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_netif/lwip/netif/wlanif.c:160
22:06:53.761 >   #4  0x401df803 in esp_netif_receive at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_netif/lwip/esp_netif_lwip.c:1259
22:06:53.761 >   #5  0x40159d79 in wifi_ap_receive at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_wifi/src/wifi_netif.c:45
22:06:53.761 >   #6  0x401979bd in hostap_deliver_data at ??:?
22:06:53.761 >   #7  0x40199b26 in hostap_input at ??:?
22:06:53.761 >   #8  0x40197b43 in ap_rx_cb at ??:?
22:06:53.761 >   #9  0x401e9ae5 in ppRxPkt at ??:?
22:06:53.761 >   #10 0x401e4c70 in ppTask at ??:?
22:06:53.761 >   #11 0x40090a02 in vPortTaskWrapper at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:162

**Reply by MicSG-dev** [reply]
> 22:17:31.205 > #3 0x40134430 in sntp_setoperatingmode at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/lwip/lwip/src/apps/sntp/sntp.c:729 (discriminator 1)

I got a similar error and other developers also got this error. It is related to NTP. See this issue: https://github.com/espressif/arduino-esp32/issues/10526

The problem has been identified. The changes that fix the problem can be seen at: https://github.com/espressif/arduino-esp32/pull/10529/files

But to make things easier, you can use the [MycilaNTP](https://github.com/mathieucarbou/MycilaNTP) library from mathieucarbou.

**Reply by mathieucarbou** [reply]
If you plan an update, point to the Arduino Core 3.1 RC2, because everybody will move to ESP-IDF 5.3 with Arduino Core 3.1 soon.

`platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip`

**Comment by juanmartin84** [comment]
thanks @mathieucarbou
